### ISSUE TITLE
Data race while performing multiple upgrades in parallel

### ISSUE THREAD

**Issue by krish7919** [helm/helm#9366]
I am using a custom Go program to perform 1000s of upgrades of helm charts in parallel.

Output of `helm version`:
From `go.mod`: `helm.sh/helm/v3 v3.5.0`

Output of `kubectl version`:
N.A. This is a unit test with memory storage/driver and `fake.PrintingKubeClient{}`.

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
Local system.

Error:

```

WARNING: DATA RACE
Write at 0x00c00033c2f0 by goroutine 45:
  helm.sh/helm/v3/pkg/action.(*Upgrade).Run()
      /Users/localuser/Desktop/workspace/example/src/example.com/release-management/release-manager/vendor/helm.sh/helm/v3/pkg/action/upgrade.go:129 +0x318
  example.com/release-management/release-manager/pkg/manager.(*helmV3Reconciler).UpgradeRelease()
      /Users/localuser/Desktop/workspace/example/src/example.com/release-management/release-manager/pkg/manager/helm3.go:315 +0x7cd
  example.com/release-management/release-manager/pkg/manager.(*helmV3Reconciler).UpgradeRelease-fm()
      /Users/localuser/Desktop/workspace/example/src/example.com/release-management/release-manager/pkg/manager/helm3.go:277 +0x4f
  example.com/release-management/release-manager/pkg/manager.(*Reconciler).asyncApply()
      /Users/localuser/Desktop/workspace/example/src/example.com/release-management/release-manager/pkg/manager/reconcile.go:263 +0x121

Previous write at 0x00c00033c2f0 by goroutine 38:
  helm.sh/helm/v3/pkg/action.(*Upgrade).Run()
      /Users/localuser/Desktop/workspace/example/src/example.com/release-management/release-manager/vendor/helm.sh/helm/v3/pkg/action/upgrade.go:129 +0x318
  example.com/release-management/release-manager/pkg/manager.(*helmV3Reconciler).UpgradeRelease()
      /Users/localuser/Desktop/workspace/example/src/example.com/release-management/release-manager/pkg/manager/helm3.go:315 +0x7cd
  example.com/release-management/release-manager/pkg/manager.(*helmV3Reconciler).UpgradeRelease-fm()
      /Users/localuser/Desktop/workspace/example/src/example.com/release-management/release-manager/pkg/manager/helm3.go:277 +0x4f
  example.com/release-management/release-manager/pkg/manager.(*Reconciler).asyncApply()
      /Users/localuser/Desktop/workspace/example/src/example.com/release-management/release-manager/pkg/manager/reconcile.go:263 +0x121

Goroutine 45 (running) created at:
  example.com/release-management/release-manager/pkg/manager.(*Reconciler).apply()
      /Users/localuser/Desktop/workspace/example/src/example.com/release-management/release-manager/pkg/manager/reconcile.go:222 +0x20e
  example.com/release-management/release-manager/pkg/manager.(*Reconciler).Reconcile()
      /Users/localuser/Desktop/workspace/example/src/example.com/release-management/release-manager/pkg/manager/reconcile.go:179 +0xdcf
  example.com/release-management/release-manager/pkg/manager.TestReconciler_Reconcile.func1()
      /Users/localuser/Desktop/workspace/example/src/example.com/release-management/release-manager/pkg/manager/reconcile_test.go:830 +0x779
  testing.tRunner()
      /usr/local/opt/go/libexec/src/testing/testing.go:1123 +0x202

Goroutine 38 (finished) created at:
  example.com/release-management/release-manager/pkg/manager.(*Reconciler).apply()
      /Users/localuser/Desktop/workspace/example/src/example.com/release-management/release-manager/pkg/manager/reconcile.go:222 +0x20e
  example.com/release-management/release-manager/pkg/manager.(*Reconciler).Reconcile()
      /Users/localuser/Desktop/workspace/example/src/example.com/release-management/release-manager/pkg/manager/reconcile.go:179 +0xdcf
  example.com/release-management/release-manager/pkg/manager.TestReconciler_Reconcile.func1()
      /Users/localuser/Desktop/workspace/example/src/example.com/release-management/release-manager/pkg/manager/reconcile_test.go:830 +0x779
  testing.tRunner()
      /usr/local/opt/go/libexec/src/testing/testing.go:1123 +0x202

```

Similar to https://github.com/helm/helm/pull/7820 and we might need a mutex to protect the update of the shared variable in https://github.com/helm/helm/blob/master/pkg/action/upgrade.go#L129: `	u.cfg.Releases.MaxHistory = u.MaxHistory`

**Comment by mattfarina** [comment]
@krish7919 A shot term fix may be to create a new action configuration for each instance of an action you run.

**Comment by krish7919** [comment]
> @krish7919 A shot term fix may be to create a new action configuration for each instance of an action you run.

I have a fix at https://github.com/krish7919/helm/tree/9366-data-race, which has been tested, I will open a PR soon.

**Comment by krish7919** [comment]
@mattfarina PR is out, please have a look, thanks!

**Comment by krish7919** [comment]
@mattfarina  Any thoughts on what can be done to fix this?

**Comment by krish7919** [comment]
> @krish7919 A shot term fix may be to create a new action configuration for each instance of an action you run.

Actually, it should be "a short term fix may be to create a new action configuration **with a new storage configuration** for each instance of an action you run.". I burnt my finger by using the in-memory storage and hit the same issue and I think I will move to a new Sql storage for my UTs to avoid this.

**Comment by krish7919** [comment]
And a new issue when using the SQL storage backend for Helm:

`panic: Unable to instantiate SQL driver: pq: sorry, too many clients already`

I see that the connection is opened by Helm at https://github.com/helm/helm/blob/master/pkg/storage/driver/sql.go#L185 but never closed ever. So when I have multiple operations in parallel, it would always run out of connections when initializing new storage configuration.

@mattfarina and team, can you guide me on how to fix this data race once and for all? This is causing a huge encumberance for us and will gladly contribute the fix to the open source community.

1. The crux of the issue seems to be [this line](https://github.com/helm/helm/blob/master/pkg/action/upgrade.go#L129) `u.cfg.Releases.MaxHistory = u.MaxHistory`, which is the only time that value is written to. In case of parallel operations and multiple instances of the the storage config being used.

2. The sql backend makes sense in never closing the connection as db connections are supposed to be long lived and do not need to be closed frequently. However, it should provide a way to close the connection in case we hit into the issue mentioned above.
